presto-main/src/main/java/com/facebook/presto/sql/planner/PlanFragmenter.java

```diff
@@ -398,6 +436,238 @@ else if (exchange.getType() == ExchangeNode.Type.REPARTITION) {
             return new RemoteSourceNode(exchange.getId(), childrenIds, exchange.getOutputSymbols(), exchange.getOrderingScheme(), exchange.getType());
         }
 
+        private PlanNode createRemoteMaterializedExchange(ExchangeNode exchange, RewriteContext<FragmentProperties> context)
+        {
+            checkArgument(exchange.getType() == REPARTITION, "Unexpected exchange type: %s", exchange.getType());
+            checkArgument(exchange.getScope() == REMOTE_MATERIALIZED, "Unexpected exchange scope: %s", exchange.getScope());
+
+            PartitioningScheme partitioningScheme = exchange.getPartitioningScheme();
+            checkArgument(!partitioningScheme.getHashColumn().isPresent(), "precomputed hashes are not supported in materializing exchanges");
+
+            PartitioningHandle partitioningHandle = partitioningScheme.getPartitioning().getHandle();
+            ConnectorId connectorId = partitioningHandle.getConnectorId()
+                    .orElseThrow(() -> new IllegalArgumentException("Unsupported partitioning handle: " + partitioningHandle));
+
+            Map<Symbol, Expression> constantExpressions = new HashMap<>();
+            Partitioning partitioning = partitioningScheme.getPartitioning();
+            List<Symbol> partitionSymbols = assignPartitioningSymbols(partitioning, constantExpressions);
+            Map<Symbol, ColumnMetadata> temporaryTableColumns = assignColumnNames(exchange.getOutputSymbols(), constantExpressions.keySet());
```

Comment:
nit: maybe just call it `symbolToColumnMap` ? Just name it `temporaryTableColumns` it's not clear it's a map (instead of a list)

Reply:
Than we should have a separate list with columns, otherwise it seems to be less readable. I would keep it as is.

Reply:
Renamed